cmd/ntpcheck/cmd/utils.go

```diff
@@ -152,40 +184,70 @@ func ntpDate(remoteServerAddr string, remoteServerPort string, requests int) err
 			return fmt.Errorf("timeout waiting for reply from server for %v", timeout)
 		}
 
+		if userDate != "" {
+			timeDelta := actualTransmitTime.Sub(clientTransmitTime)
+			clientReceiveTime = clientReceiveTime.Add(-timeDelta)
+		}
+
 		serverReceiveTime := ntp.Unix(response.RxTimeSec, response.RxTimeFrac)
 		serverTransmitTime := ntp.Unix(response.TxTimeSec, response.TxTimeFrac)
-		clientOriginTime := ntp.Unix(response.OrigTimeSec, response.OrigTimeFrac)
+		originTime := ntp.Unix(response.OrigTimeSec, response.OrigTimeFrac)
 
-		log.Debugf("Client TX timestamp: %v, Origin TX timestamp: %v", clientTransmitTime, clientOriginTime)
-		log.Debugf("Server RX timestamp: %v", serverReceiveTime)
-		log.Debugf("Server TX timestamp: %v", serverTransmitTime)
-		log.Debugf("Client RX timestamp: %v", clientReceiveTime)
+		log.Debugf("Client TX timestamp: %v, Origin TX timestamp (T1): %v", clientWireTransmitTime, originTime)
+		log.Debugf("Server RX timestamp (T2): %v", serverReceiveTime)
+		log.Debugf("Server TX timestamp (T3): %v", serverTransmitTime)
+		log.Debugf("Client RX timestamp (T4): %v", clientReceiveTime)
 
-		// sanity check: origin time must be same as our transmit time
+		// sanity check: origin time must be same as client transmit time
 		if response.OrigTimeSec != sec || response.OrigTimeFrac != frac {
-			log.Errorf("Client TX timestamp %v not equal to Origin TX timestamp %v", clientTransmitTime, clientOriginTime)
+			failure += 1
+			log.Errorf("Client TX timestamp %v not equal to Origin TX timestamp %v", clientTransmitTime, originTime)
+		} else {
+			success += 1
 		}
 
-		avgNetworkDelay := ntp.AvgNetworkDelay(clientOriginTime, serverReceiveTime, serverTransmitTime, clientReceiveTime)
-		currentRealTime := ntp.CurrentRealTime(serverTransmitTime, avgNetworkDelay)
-		offset := ntp.CalculateOffset(currentRealTime, time.Now())
+		delay := ntp.RoundTripDelay(originTime, serverReceiveTime, serverTransmitTime, clientReceiveTime)
+		offset := ntp.Offset(originTime, serverReceiveTime, serverTransmitTime, clientReceiveTime)
+		correctTime := ntp.CorrectTime(clientReceiveTime, offset)
 
-		sumAvgNetworkDelay += avgNetworkDelay
+		sumDelay += delay
 		sumOffset += offset
 
-		// On last request calculate everything
 		if i == requests-1 {
-			fmt.Printf("Last:\n")
-			fmt.Printf("Stratum: %d, Current time: %s\n", response.Stratum, currentRealTime)
-			fmt.Printf("Offset: %fs (%fms), Network delay: %fs (%fms)\n", float64(offset)/float64(time.Second.Nanoseconds()), float64(offset)/float64(time.Millisecond.Nanoseconds()), float64(avgNetworkDelay)/float64(time.Second.Nanoseconds()), float64(avgNetworkDelay)/float64(time.Millisecond.Nanoseconds()))
+			fmt.Printf("\nServer: %s (%s), Stratum: %d\n", addr, toByteArray(response.ReferenceID), response.Stratum)
+			fmt.Printf("*** %d NTP client packets sent: successful %d / failed %d ***\n", requests, success, failure)
+		}
+		if requests == 1 {
```

Comment:
Why do we do it only when requests is 1?
And if we don't check that flags (-x and 1 request) are incompatible before we run this, it's a bug, as no explanation will be given despite the -x flag.

If we really do want this logic, let's at least combine the condition to avoid double indentation: `if requests == 1 && explain {`

Reply:
Added logic to report to user that `-x` and multiple requests are incompatible.